### DISCUSSION TITLE
utilize global dictionaries to improve partitioning / sorting performance for variable-width types

### DISCUSSION THREAD

**Discussion by mhseiden** [duckdb/duckdb#6800] [Ideas]
# Overview
**TL;DR using global dictionaries for partitioning / aggregation / join / sort keys and an approach to external sort that does not transpose+spill the full table could improve performance and reliability of DuckDB.**

I had a hypothesis was that using fixed-width integers would improve the runtime of partition/sort operations, as well as reduce the volume of data spilled to disk. I also assumed the dictionary encode / decode overhead is also likely negligible for sufficiently small dictionaries, while construction is a function of base data size. 

To validate this hypothesis I've run experiments on an AWS EC2 `c7g.large` instance -- ARM Gravitron3, 2vCPU, 4GB memory, ~130MB/s reads w/ 8GB SSDs using TPC-H SF1. I've found the resource constraints useful in these tests as they stress DuckDB's ability to yield good reliable performance under non-ideal circumstances. Through the experimentation I've developed SQL statements that performs global dictionary construction+encoding+decoding of low-cardinality string columns, and which improve execution time of specific windowed calculations by almost 2x. There's certainly a lot more to investigate around the approach, but I wanted to share it here as it seems like an effective technique DuckDB could employ internally. 

Below are the two tests I performed. One computes a row number over a narrow projection of the sf1 lineitem table while the other adds the additional column to the full projection. Both are reductions of actual queries produced by our product at Sigma Computing, where I work.

# Improving Case 1
Case 1 assigns row numbers to all records within a partition, where partitions are defined by a multi-column key with a low number of distinct groups. The final projection is narrow, and contains the computed row number and the two partition by columns.

## ~2x improvement w/ default configuration
Here is the baseline SQL statement I set out to improve.
```sql
select l_shipinstruct, l_shipmode, row_number() over (partition by l_shipinstruct, l_shipmode) from lineitem
```
With default settings the query reliably completes with with approximately `Total Time: 1.54s`. 

Here is the SQL statement I developed that performs dictionary encoding.
```sql
with 
dict_l_shipinstruct as (
	select l_shipinstruct, (row_number() over (order by l_shipinstruct))::TINYINT id 
	  from (select distinct l_shipinstruct from lineitem)),

dict_l_shipmode as (
	select l_shipmode, (row_number() over (order by l_shipmode))::TINYINT id 
	  from (select distinct l_shipmode from lineitem)),

encode as (
	select dict_l_shipinstruct.id l_shipinstruct_id, dict_l_shipmode.id l_shipmode_id 
	  from lineitem 
 	  join dict_l_shipmode using (l_shipmode) 
	  join dict_l_shipinstruct using (l_shipinstruct)),

rn as (
	select l_shipinstruct_id, l_shipmode_id, row_number() over (partition by l_shipinstruct_id, l_shipmode_id) rn 
	from encode)

select l_shipinstruct, l_shipmode, rn.rn row_number
from rn
join dict_l_shipinstruct on l_shipinstruct_id = dict_l_shipinstruct.id
join dict_l_shipmode on l_shipmode_id = dict_l_shipmode.id
```
With default settings the query reliably completes with approximately `Total Time: 0.849s`

## Improvements with low memory limits.
When I restrict the memory limit to `1gb` the performance is roughly the same. The dictionary encoding runtime stays the same. However the baseline SQL runtime does increase to roughly `Total Time: 1.77s` which is ~10% slower. This is measurable, but not hugely impactful in a practical sense.

When I restrict the memory limit to `512mb` the performance gap widens. The dictionary encoding runtime again stays the same. However the baseline SQL increases to roughly `Total Time: 2.98s`. This is almost a 3x gap in performance.

When I restrict the memory limit to `256mb` the experiment fails. The dictionary encoded SQL fails with an OOM, likely because joins don't yet spill to disk. Meanwhile the baseline SQL increases to between `5s` and `6s`, likely driven by spilling in the window operator.

## Takeaway
The dictionary encoding is able to remain resident in memory longer, which is likely why the performance stays flat as memory drops. Once spilling is implemented for joins it'd be interesting to re-run the experiment to see how the two out-of-core approaches compare. I'd still expect the dictionary encoded approach to perform better since it would spill less data overall, however it'd be nice to confirm this experimentally.

# Improving Case 2
Case 2 assigns row numbers to all records within a partition, where partitions are defined by a multi-column key with a low number of distinct groups. The final projection includes the full table and the the computed row number.

Here is the baseline SQL statement I set out to improve.
```sql
select *, row_number() over (partition by l_shipinstruct, l_shipmode) from lineitem
```
With default settings the query reliably completes with with approximately `Total Time: 3.38s`. 

Here is the SQL statement I developed that performs dictionary encoding.
```sql
with 
dict_l_shipinstruct as (
	select l_shipinstruct, (row_number() over (order by l_shipinstruct))::TINYINT id 
	  from (select distinct l_shipinstruct from lineitem)),

dict_l_shipmode as (
	select l_shipmode, (row_number() over (order by l_shipmode))::TINYINT id 
	  from (select distinct l_shipmode from lineitem)),

global_rid as (
    select *, row_number() over () grid
    from lineitem),

encode as (
	select dict_l_shipinstruct.id l_shipinstruct_id, dict_l_shipmode.id l_shipmode_id, grid
	  from global_rid 
 	  join dict_l_shipmode using (l_shipmode) 
	  join dict_l_shipinstruct using (l_shipinstruct)),

rn as (
	select grid, row_number() over (partition by l_shipinstruct_id, l_shipmode_id) rn 
	from encode)

select global_rid.*, rn.rn row_number
from global_rid
join rn using (grid)
```

With default settings the query reliably completes with approximately `Total Time: 2.28s`

The major difference between the two approaches is the amount of data that the window operator would spill. Rather than transposing and spilling the full record to disk, it would (in theory) spill the minimal data necessary to execute the window, and assign each row the computed value via an integer-key join. More generally, this technique could aid external sort by avoiding shuffling the whole table to and from disk, and instead producing a sort vector that's joined back to the wide table.

## Improvements with low memory limits.
When I restrict the memory limit to `1gb` the performance gap widens drastically. The dictionary encoding runtime stays roughly the same. However the baseline SQL runtime increases to `Total Time: 16.26s` -- that's almost an 8x increase in runtime.

When I restrict the memory limit to `512mb` the test fails. The dictionary encoding runtime again stays roughly the same. However the baseline SQL query fails to complete because spilling consumes all space on the disk. 

## Takeaways
At face value, the DuckDB's external sorting implementation is not well suited for "wide tables", due to the transposition and spilling of the full table data. The combination of dictionary encoding and global row identifiers enables both lower memory consumption and the ability to spill only the data needed to produce a sort vector, which could be orders of magnitude less than the full table.

## A note on determinism w/ CTEs
Once concern I have about the correctness of the approach is that `row_number() over ()` does not produce deterministic results if performed multiple times in the query plan. I believe this could be an issue as (from what I can tell) the intermediate results from CTEs are not written to temp space when referenced +2 times in the overall statement. However, I don't believe this has any impact the performance characteristics, beyond the incremental cost of deterministic global row id generation. Since DuckDB is a single process on a single node it seems that this global numbering can (theoretically) be done cheaply.

# Discussion
My primary takeaway is that using global dictionaries and an approach to external sort that does not transpose+spill the full table could improve performance and reliability of DuckDB. My SQL-based experiments above indicate this approach holds some promise. Given that Arrow supports per-chunk dictionaries it seems plausible to extend that compression technique to produce global dictionaries. The use of dictionaries (construction+encoding+decoding) could then factor into query optimization as physical plan alternatives when costing equivalent plans. In my example above this technique is applied to windows (w/ external sorts) but it could also apply to joins and aggregations. Using integer keys would speedup hash code calculations, and shrink the in-memory size of hashtables.

**Comment by Mytherin** [comment]
Thanks for the detailed analysis - making use of dictionaries in various operations (particularly grouping and partitioning) is definitely on our roadmap. Especially as the dictionaries from our storage can be directly used in the various operators, this could lead to large performance gains as you have correctly seen.  Of course the TPC-H data is an extreme example, since the columns only have on the order of 2-6 distinct values, there are definitely large potential gains to be had here on many data sets.

**Reply by mhseiden** [reply]
That's great to hear this is on the roadmap. Looking forward to taking it for a spin, when it lands.
